File: advent_2015/src/python/advent_2015_day_04.py
import hashlib
input_str = "yzbqklnj"

def hash_till(starts_with_prefix: str) -> str:
    suffix = 1
    while(True):
        output_hash = hashlib.md5((input_str + str(suffix)).encode('utf-8')).hexdigest()
        if output_hash.startswith(starts_with_prefix):
            return f"hash: {output_hash}, suffix: {suffix}"
        suffix+=1

# part1
print(hash_till("00000"))

# part2
print(hash_till("000000"))

File: advent_2015/src/bin/advent_2015_day_04.rs
use md5;

fn part_1(input_str: &str) -> usize {
    let mut result = 1;
    for suffix in 1.. {
        let digest = md5::compute(format!("{input_str}{suffix}"));
        if digest[..2] == [0, 0] && digest[2] <= 0x0F {
            result = suffix;
            break;
        }
    }
    result
}

fn part_2(input_str: &str) -> usize {
    let mut result = 1;
    for suffix in 1.. {
        let digest = md5::compute(format!("{input_str}{suffix}"));
        if digest[..3] == [0, 0, 0] {
            result = suffix;
            break;
        }
    }
    result
}

fn main() {
    let input_str: &str = "yzbqklnj";
    println!("{}", part_1(input_str));
    println!("{}", part_2(input_str));
}
